grpc/src/main/java/io/stargate/grpc/server/interceptors/AuthenticationInterceptor.java

```diff
@@ -0,0 +1,43 @@
+package io.stargate.grpc.server.interceptors;
+
+import io.grpc.Context;
+import io.grpc.Contexts;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+import io.stargate.auth.AuthenticationService;
+import io.stargate.auth.UnauthorizedException;
+import io.stargate.grpc.server.Server;
+
+public class AuthenticationInterceptor implements ServerInterceptor {
+  public static final Metadata.Key<String> TOKEN_KEY =
+      Metadata.Key.of("X-Cassandra-Token", Metadata.ASCII_STRING_MARSHALLER);
+
+  private final AuthenticationService authentication;
+
+  public AuthenticationInterceptor(AuthenticationService authentication) {
+    this.authentication = authentication;
+  }
+
+  @Override
+  public <ReqT, RespT> Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+    try {
+      String token = headers.get(TOKEN_KEY);
+      if (token == null) {
+        call.close(Status.UNAUTHENTICATED.withDescription("No token provided"), new Metadata());
+        return null;
+      }
+      Context context = Context.current();
+      context = context.withValue(Server.AUTHENTICATION_KEY, authentication.validateToken(token));
```

Comment:
Maybe too early for this, but we're going to have to provide something like `ClientInfo` in this context for parity with authentication over CQL.

Also, CQL authentication goes through `SaslNegotiator`, which does not seem to be relevant here. So, we may need to add a new method to `AuthenticationService` for properly handling non-HTTP connections without SASL.

Reply:
Having thought more about this, I believe it is preferable to call `validateToken(token, ClientInfo))`. The single parameter method is not guaranteed to be working in all environments as its input data is not always sufficient for authentication.

This is probably ok for now, but we should fix before we consider gRPC for use in the cloud.

Reply:
Good call. I'm not sure how to handle that yet... I'm not sure the way we're handling that with CQL is going to work for gRPC. I'll add a JIRA to fix that for cloud.